coderd/devtunnel/tunnel.go

```diff
@@ -1,93 +1,270 @@
+	"golang.zx2c4.com/wireguard/wgctrl/wgtypes"
+
+	"cdr.dev/slog"
+)
+
+const (
+	EndpointWireguard = "wg-tunnel-udp.coder.app"
+	EndpointHTTPS     = "wg-tunnel.coder.app"
 )
 
-// New creates a new tunnel pointing at the URL provided. Once created, it
-// returns the external hostname that will resolve to it.
-//
-// The tunnel will exit when the context provided is canceled.
-//
-// Upstream connection occurs synchronously through a selfhosted
-// https://github.com/fatedier/frp instance. The error channel sends an error
-// when the frp client stops.
-func New(ctx context.Context, coderurl *url.URL) (string, <-chan error, error) {
-	frpcrypto.DefaultSalt = "frp"
-
-	cfg := frpconfig.GetDefaultClientConf()
-	cfg.ServerAddr = "frp-tunnel.coder.app"
-	cfg.ServerPort = 7000
-
-	// Ignore all logs from frp.
-	frplog.InitLog("file", os.DevNull, "error", -1, false)
-
-	var (
-		id        = uuid.NewString()
-		subdomain = strings.ReplaceAll(namesgenerator.GetRandomName(1), "_", "-")
-		portStr   = coderurl.Port()
+type Tunnel struct {
+	URL       string
+	Listener  net.Listener
+	ErrorChan <-chan error
+}
+
+type Config struct {
+	ID         uuid.UUID              `json:"id"`
+	PrivateKey device.NoisePrivateKey `json:"private_key"`
+	PublicKey  device.NoisePublicKey  `json:"public_key"`
+}
+type configExt struct {
+	ID         uuid.UUID              `json:"id"`
+	PrivateKey device.NoisePrivateKey `json:"-"`
+	PublicKey  device.NoisePublicKey  `json:"public_key"`
+}
+
+// NewWithConfig calls New with the given config. For documentation, see New.
+func NewWithConfig(ctx context.Context, logger slog.Logger, cfg Config) (*Tunnel, error) {
+	err := startUpdateRoutine(ctx, logger, cfg)
+	if err != nil {
+		return nil, xerrors.Errorf("start update routine: %w", err)
+	}
+
+	tun, tnet, err := netstack.CreateNetTUN(
+		[]netip.Addr{netip.AddrFrom16(cfg.ID)},
+		[]netip.Addr{netip.AddrFrom4([4]byte{1, 1, 1, 1})},
+		1420,
 	)
-	if portStr == "" {
-		portStr = "80"
+	if err != nil {
+		return nil, xerrors.Errorf("create net TUN: %w", err)
 	}
 
-	port, err := strconv.ParseInt(portStr, 10, 64)
+	wgip, err := net.ResolveIPAddr("ip", EndpointWireguard)
 	if err != nil {
-		return "", nil, xerrors.Errorf("parse port %q: %w", port, err)
+		return nil, xerrors.Errorf("resolve endpoint: %w", err)
 	}
 
-	httpcfg := map[string]frpconfig.ProxyConf{
-		id: &frpconfig.HTTPProxyConf{
-			BaseProxyConf: frpconfig.BaseProxyConf{
-				ProxyName:      id,
-				ProxyType:      frpconsts.HTTPProxy,
-				UseEncryption:  false,
-				UseCompression: false,
-				LocalSvrConf: frpconfig.LocalSvrConf{
-					LocalIP:   coderurl.Hostname(),
-					LocalPort: int(port),
-				},
-			},
-			DomainConf: frpconfig.DomainConf{
-				SubDomain: subdomain,
-			},
-			Locations: []string{""},
-		},
+	dev := device.NewDevice(tun, conn.NewDefaultBind(), device.NewLogger(device.LogLevelVerbose, ""))
+	err = dev.IpcSet(fmt.Sprintf(`private_key=%s
+public_key=%s
+endpoint=%s:55555
```

Comment:
Is `55555` a static port we use?

Reply:
Yeah, that's the UDP port on the server for Wireguard to connect to.